controllers/machine_controller.go

```diff
@@ -321,6 +322,20 @@ func (r *MachineReconciler) reconcileDelete(ctx context.Context, cluster *cluste
 
 			conditions.MarkTrue(m, clusterv1.DrainingSucceededCondition)
 			r.recorder.Eventf(m, corev1.EventTypeNormal, "SuccessfulDrainNode", "success draining Machine's node %q", m.Status.NodeRef.Name)
+
+			// after node draining, make sure volumes are detached before delete node
+			if conditions.Get(m, clusterv1.VolumeDetachSucceededCondition) == nil {
+				conditions.MarkFalse(m, clusterv1.VolumeDetachSucceededCondition, clusterv1.WaitingForVolumeDetachReason, clusterv1.ConditionSeverityInfo, "Waiting for node volumes to be detached")
+			}
+			if ok, err := r.isNodeVolumesDetached(ctx, cluster, m.Status.NodeRef.Name, m.Name); !ok || err != nil {
+				if err != nil {
+					r.recorder.Eventf(m, corev1.EventTypeWarning, "FailedWaitForVolumeDetach", "error wait for volume detach, node %q: %v", m.Status.NodeRef.Name, err)
+					return ctrl.Result{}, err
+				}
+				return ctrl.Result{}, nil
```

Comment:
Should we log something here that informs the user we're waiting for volume detachment?